File: smoke_tests/php/unsafe_reflection.php
<?php
// Unsafe Reflection vulnerabilities in PHP

// Test 1: Dynamic class instantiation
function load_class() {
    $class_name = $_GET['class'];
    // VULNERABLE: User controls class instantiation
    $instance = new $class_name();
    return $instance;
}

// Test 2: ReflectionClass with user input
function reflect_class() {
    $class_name = $_GET['class'];
    // VULNERABLE: User controls reflection
    $reflection = new ReflectionClass($class_name);
    return $reflection->getMethods();
}

// Test 3: Dynamic method call
function invoke_method() {
    $class = $_POST['class'];
    $method = $_POST['method'];

    $instance = new $class();
    // VULNERABLE: User controls method invocation
    return $instance->$method();
}

// Test 4: call_user_func with user input
function call_function() {
    $callback = $_GET['callback'];
    $args = $_GET['args'] ?? [];
    // VULNERABLE: User controls function call
    return call_user_func($callback, ...$args);
}

// Test 5: call_user_func_array
function call_function_array() {
    $function = $_POST['function'];
    $arguments = $_POST['arguments'];
    // VULNERABLE: User controls function and args
    return call_user_func_array($function, $arguments);
}

// Test 6: ReflectionMethod invoke
function reflect_invoke() {
    $class = $_GET['class'];
    $method = $_GET['method'];

    $instance = new $class();
    $reflection = new ReflectionMethod($class, $method);
    // VULNERABLE: Invoking user-specified method
    return $reflection->invoke($instance);
}

// Test 7: Property access by name
function get_property() {
    $class = $_GET['class'];
    $property = $_GET['property'];

    $instance = new $class();
    // VULNERABLE: User controls property access
    return $instance->$property;
}

// Test 8: ReflectionProperty access
function reflect_property() {
    $class = $_GET['class'];
    $property = $_GET['property'];

    $instance = new $class();
    $reflection = new ReflectionProperty($class, $property);
    $reflection->setAccessible(true);
    // VULNERABLE: Accessing private properties
    return $reflection->getValue($instance);
}

// Test 9: class_exists check bypass
function create_instance() {
    $class = $_GET['class'];
    // VULNERABLE: class_exists doesn't prevent malicious instantiation
    if (class_exists($class)) {
        return new $class();
    }
}

// Test 10: Forward static call
function static_call() {
    $class = $_POST['class'];
    $method = $_POST['method'];
    $args = $_POST['args'] ?? [];
    // VULNERABLE: User controls static method call
    return forward_static_call([$class, $method], ...$args);
}

// Test 11: Variable function call
function variable_function() {
    $func = $_GET['func'];
    // VULNERABLE: User controls function execution
    return $func();
}

// Test 12: Create function (deprecated but still risky)
function create_func() {
    $code = $_POST['code'];
    // VULNERABLE: Code execution
    // $func = create_function('$x', $code);  // Deprecated
    // return $func(1);
}

// Test 13: Instantiate with constructor args
function create_with_args() {
    $class = $_GET['class'];
    $args = $_GET['args'] ?? [];
    // VULNERABLE: User controls class and constructor
    $reflection = new ReflectionClass($class);
    return $reflection->newInstanceArgs($args);
}
?>

File: smoke_tests/csharp/unsafe_reflection.cs
// Unsafe Reflection vulnerabilities in C#
using System;
using System.Reflection;
using System.Web.Mvc;

namespace VulnerableApp
{
    public class ReflectionController : Controller
    {
        // Test 1: Type.GetType with user input
        public ActionResult LoadType()
        {
            string typeName = Request.QueryString["type"];
            // VULNERABLE: User controls type instantiation
            Type type = Type.GetType(typeName);
            if (type != null)
            {
                object instance = Activator.CreateInstance(type);
                return Json(instance.ToString());
            }
            return NotFound();
        }

        // Test 2: Assembly.Load with user input
        public ActionResult LoadAssembly()
        {
            string assemblyName = Request.QueryString["assembly"];
            // VULNERABLE: User can load arbitrary assemblies
            Assembly assembly = Assembly.Load(assemblyName);
            return Json(assembly.GetTypes().Length);
        }

        // Test 3: Method invocation by name
        public ActionResult InvokeMethod()
        {
            string typeName = Request.Form["type"];
            string methodName = Request.Form["method"];

            Type type = Type.GetType(typeName);
            // VULNERABLE: User controls method invocation
            MethodInfo method = type?.GetMethod(methodName);
            object instance = Activator.CreateInstance(type);
            object result = method?.Invoke(instance, null);
            return Json(result);
        }

        // Test 4: Property access by name
        public ActionResult GetProperty()
        {
            string typeName = Request.QueryString["type"];
            string propName = Request.QueryString["property"];

            Type type = Type.GetType(typeName);
            object instance = Activator.CreateInstance(type);
            // VULNERABLE: User-controlled property access
            PropertyInfo prop = type?.GetProperty(propName);
            object value = prop?.GetValue(instance);
            return Json(value);
        }

        // Test 5: Field access by name
        public ActionResult GetField()
        {
            string typeName = Request.QueryString["type"];
            string fieldName = Request.QueryString["field"];

            Type type = Type.GetType(typeName);
            object instance = Activator.CreateInstance(type);
            // VULNERABLE: User-controlled field access
            FieldInfo field = type?.GetField(fieldName,
                BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
            return Json(field?.GetValue(instance));
        }

        // Test 6: Constructor invocation with parameters
        public ActionResult CreateWithParams()
        {
            string typeName = Request.Form["type"];
            string param = Request.Form["param"];

            Type type = Type.GetType(typeName);
            // VULNERABLE: User-controlled instantiation
            object instance = Activator.CreateInstance(type, param);
            return Json(instance.ToString());
        }

        // Test 7: Assembly.LoadFrom with path
        public ActionResult LoadFromPath()
        {
            string path = Request.QueryString["path"];
            // VULNERABLE: Load assembly from user-controlled path
            Assembly assembly = Assembly.LoadFrom(path);
            return Json(assembly.FullName);
        }

        // Test 8: Dynamic method invocation with arguments
        public ActionResult InvokeWithArgs()
        {
            string typeName = Request.Form["type"];
            string methodName = Request.Form["method"];
            string[] args = Request.Form["args"]?.Split(',');

            Type type = Type.GetType(typeName);
            object instance = Activator.CreateInstance(type);
            MethodInfo method = type?.GetMethod(methodName);
            // VULNERABLE: User controls method and arguments
            object result = method?.Invoke(instance, args);
            return Json(result);
        }

        // Test 9: Setting property value
        public ActionResult SetProperty()
        {
            string typeName = Request.Form["type"];
            string propName = Request.Form["property"];
            string value = Request.Form["value"];

            Type type = Type.GetType(typeName);
            object instance = Activator.CreateInstance(type);
            PropertyInfo prop = type?.GetProperty(propName);
            // VULNERABLE: User-controlled property modification
            prop?.SetValue(instance, value);
            return Ok();
        }

        // Test 10: GetMethod with binding flags
        public ActionResult InvokePrivate()
        {
            string typeName = Request.Form["type"];
            string methodName = Request.Form["method"];

            Type type = Type.GetType(typeName);
            object instance = Activator.CreateInstance(type);
            // VULNERABLE: Accessing private methods
            MethodInfo method = type?.GetMethod(methodName,
                BindingFlags.NonPublic | BindingFlags.Instance);
            return Json(method?.Invoke(instance, null));
        }
    }
}
